birdhouse/FreeCad.py

```diff
@@ -0,0 +1,140 @@
+        ActiveSketch.addConstraint(Sketcher.Constraint('Coincident',lineIdx,1,pointIdx1,1))
+        ActiveSketch.addConstraint(Sketcher.Constraint('Coincident',lineIdx,2,pointIdx2,1))
+        return lineIdx
+
+    myMap = {}
+    for index in range(len(points)):
+        xPoint = 0.5 if index == 0 else index
+        yPoint = index
+        pointIdx = ActiveSketch.addGeometry(Part.Point(App.Vector(xPoint,yPoint,0)))
+        myMap[points[index]['name']] = pointIdx
+        points[index]['index'] = pointIdx
+
+
+    lineMap = {}
+    for index in range(len(points)):
+        previousIndex = len(points) - 1 if index == 0 else index - 1
+        previousPointIndex = points[previousIndex]['index']
+        currentPointIndex = points[index]['index']
+        lineIdx = addLine(previousPointIndex, currentPointIndex)
+        previousPointName = points[previousIndex]['name']
+        currentPointName = points[index]['name']
+        lineMap["{previousPointName}-{currentPointName}".format(previousPointName=previousPointName, currentPointName=currentPointName)] = lineIdx
+
+    def findLineIndex(lineName = ""):
+        if lineName is None:
+            return None
+        nameParts = lineName.split("-")
+        alternateName = nameParts[1] + '-' + nameParts[0]
+        return lineMap.get(lineName) or lineMap.get(alternateName)
+
+    for constraint in constraints:
+        line = constraint.get('line')
+        point = constraint.get('point')
+        conType = constraint['type']
+        lineIdx = findLineIndex(line)
+        pointIdx = myMap.get(point)
+        if conType == 'Horizontal':
+            idx = ActiveSketch.addConstraint(Sketcher.Constraint('Horizontal',lineIdx))
+        elif conType == 'Distance':
+            distance = constraint['value']
+            ActiveSketch.addConstraint(Sketcher.Constraint('Distance',lineIdx,distance))
+        elif conType == 'OnXAxis':
+            ActiveSketch.addConstraint(Sketcher.Constraint('PointOnObject',lineIdx,1,-1))
+        elif conType == 'OriginXDistance':
+            xAxisIdx = -1
+            distance = constraint['value']
+            ActiveSketch.addConstraint(Sketcher.Constraint('DistanceX',pointIdx,1,xAxisIdx,1,distance))
+        elif conType == 'OnYAxis':
+            yAxisIdx = -2
+            ActiveSketch.addConstraint(Sketcher.Constraint('PointOnObject',pointIdx,1,yAxisIdx))
+        elif conType == 'perpendicularDistance':
+            distance = constraint['value']
+            ActiveSketch.addConstraint(Sketcher.Constraint('Distance',pointIdx,1,lineIdx,distance))
+
+def makeFrame(partName, plane = 'YZ_Plane'):
+    points = [
+        {
+            'name': 'baseR',
+        },
+        {
+            'name': 'baseL',
+        },
+        {
+            'name': 'top',
+        },
+    ]
+
+    constraints = [
+        {
+            'type': 'Horizontal',
+            'line': 'baseR-baseL'
```

Comment:
Here the line made up of the points `baseR` and `baseL` is constrained to be horizontal, and so and so forth with the rest of the constraints. 
I didn't put a great deal of thought into coming up the perfect API for this, as just trying to get a proof of concept working.